### DISCUSSION TITLE
Can we use API to generate invite token?

### DISCUSSION THREAD

**Discussion by LideDing** [gravitational/teleport#33114] [Q&A] [answered]
Hello, I want to know if there is a way to generate invite token by API.
For example,the command `tctl nodes add --ttl=2h --roles=node` has such ouput:
```
The invite token: 3c7b8b102f3eec8d34d9eb0a2731e467
This token will expire in 120 minutes.

Run this on the new node to join the cluster:

> teleport start \
   --roles=node \
   --token=3c7b8b102f3eec8d34d9eb0a2731e467 \
   --ca-pin=sha256:6d1c9a7fe3113b7569faef83b2acc192dc23e96e9af1a882ec70a6b0c7e73325 \
   --auth-server=10.2.9.22:3025

Please note:

  - This invitation token will expire in 120 minutes
  - 10.2.9.22:3025 must be reachable from the new node
 ```
Now I want to use Teleport API to get the token, for some reasons I can't use CLI. Does anyone know how to solve this problem? Thanks a lot!!

**Comment by LideDing** [comment]
I think I know about it! The function `CreateToken` is not for generating token, but for token insert.

**Reply by webvictim** [reply, marked as answer]
Take a look at https://pkg.go.dev/github.com/gravitational/teleport/api/client#Client.UpsertToken - that's probably what you want.

Teleport's token generation logic is in the client itself rather than handled by the API - if you want to generate random tokens, you should do that in your own client and then use `CreateToken` or `UpsertToken` to add it to the backend.

**Reply by LideDing** [reply]
Thank you so much! I've made it successfully. And I noticed that before I use go API, I have to use `tsh login` to login manually. I can not use any CLI for my project, I think maybe machineID can solve this issue but I can't find a referrnce. Is there any example code or reference I can find online? Thank you!

**Reply by webvictim** [reply]
Our Machine ID docs were recently overhauled! Take a look at this guide on how to use it with `tctl`: https://goteleport.com/docs/machine-id/access-guides/tctl/

This should give you a foundation of how to use Machine ID to get an identity file, which you can then use with your API client without needing `tsh login`.

**Reply by LideDing** [reply]
Thanks! It helps a lot
